hedera-node/hedera-app/src/main/java/com/hedera/node/app/ServicesMain.java

```diff
@@ -321,6 +353,118 @@ public static void main(final String... args) throws Exception {
         hedera.run();
     }
 
+    /**
+     * Determining the Roster History.
+     * There are three non-genesis modes that a node can start in:
+     * <ul>
+     *   <li> Genesis Network - The node is started with a genesis roster and no pre-existing state on disk. </li>
+     *   <li> Network Transplant - The node is started with a state on disk and an overriding roster for a different network. </li>
+     *   <li> Software Upgrade - The node is restarted with the same state on disk and a software upgrade is happening. </li>
+     *   <li> Normal Restart - The node is restarted with the same state on disk and no software upgrade is happening. </li>
+     * </ul>
+     *
+     * @param isGenesis            whether running in genesis mode
+     * @param state                the state of the platform
+     * @param rosterStore          the roster store
+     * @param diskAddressBook     the address book read from disk
+     * @param softwareUpgrade      whether a software upgrade is happening
+     * @return the roster history
+     */
+    private static @NonNull RosterHistory determineRosterHistory(
+            final boolean isGenesis,
+            @NonNull final State state,
+            @NonNull final WritableRosterStore rosterStore,
+            @NonNull final AddressBook diskAddressBook,
+            final boolean softwareUpgrade) {
+        requireNonNull(state);
+        requireNonNull(rosterStore);
+        requireNonNull(diskAddressBook);
+
+        if (isGenesis) {
+            final var genesisRoster = loadRoster(GENESIS_CONFIG_FILE_NAME);
+            // Set (genesisRoster, 0) ase the new active roster in the roster state.
+            rosterStore.putActiveRoster(genesisRoster, 0);
+
+            // rosterHistory := [(genesisRoster, 0)]
+            return new RosterHistory(genesisRoster, 0, genesisRoster, 0);
+        }
+
+        final boolean overrideConfigExists = false; // An override-config.txt file is present on disk
+        if (overrideConfigExists) {
+            // FUTURE: Network Transplant
+        }
+
+        // Normal Restart, no software upgrade is happening
+        if (!softwareUpgrade) {
```

Comment:
The Roster choosing logic can be in a separate class

Reply:
Agree. Also, each section of logic can be extracted to a separate method with an informative name.

Reply:
Added `RosterStartupLogic` class.